internal/pkg/crawl/extractor/xml.go

```diff
@@ -19,25 +23,39 @@ func XML(resp *http.Response) (URLs []*url.URL, sitemap bool, err error) {
 		sitemap = true
 	}
 
-	mv, err := mxj.NewMapXml(xmlBody)
-	if err != nil {
-		return nil, sitemap, err
-	}
+	decoder := xml.NewDecoder(strings.NewReader(string(xmlBody)))
+	var (
+		startElement xml.StartElement
+		currentNode  *LeafNode
+		leafNodes    []LeafNode
+	)
 
-	// Try to find if it's a sitemap
-	for _, node := range mv.LeafNodes() {
-		if strings.Contains(node.Path, "sitemap") {
-			sitemap = true
+	for {
+		tok, err := decoder.Token()
+		if err == io.EOF {
 			break
 		}
-	}
+		if err != nil {
+			return nil, sitemap, err
+		}
 
-	for _, value := range mv.LeafValues() {
-		if _, ok := value.(string); ok {
-			if strings.HasPrefix(value.(string), "http") {
-				URL, err := url.Parse(value.(string))
+		switch tok := tok.(type) {
+		case xml.StartElement:
```

Comment:
Right, opening tags also have urls in some cases.

Now the only left is `TestXMLBodyReadError`, gotta look at it

Reply:
I think that test is invalid(?) since `NopCloser` certainly won't return an EOF error on `xmlBody, err := io.ReadAll(resp.Body)`

Reply:
Possible, but they did pass previously. Sure, it does not return an EOF on read. For the test to pass, I had to decode the Token once and catch the error. (and seek back for the loop)

```go
_, err = decoder.Token()
if err != nil {
	return nil, sitemap, err
}

// seek back to 0 if we are still here
reader.Seek(0, 0)
decoder = xml.NewDecoder(reader)
```

Catching this in the loop won't work cleanly, since I want to know if EOF was somewhere in-between the file (invalid XML), or at the start (this error)

I will push these changes

Reply:
If EOF was somewhere in-between the file (invalid XML), the original EOF error will be replaced by a `Decoder.syntaxError("unexpected EOF")` error.

https://github.com/golang/go/blob/2b33434287be85917ba8edc4a73f0a6ebaaea63e/src/encoding/xml/xml.go#L285-L288